### ISSUE TITLE
PolyLineWidget: lines are incorrectly rendered

### ISSUE THREAD

**Issue by floryst** [Kitware/vtk-js#1815]
**Describe the bug**

The PolyLineWidget lines are incorrectly rendered, as seen below.

![image](https://user-images.githubusercontent.com/1812167/113218032-5da3b600-926e-11eb-94ec-f724653142e5.png)

I have not yet done any digging into the cause of this issue.

**To Reproduce**

1. https://kitware.github.io/vtk-js/examples/PolyLineWidget.html
2. Grab Focus, and start placing points.

**Expected behavior**

Lines should be connecting each handle point.

**Platform:**

This was detected on Chrome and Firefox for Linux.

***vtk.js version:***
v17.2.2

**Comment by finetjul** [comment]
I wonder if #1810  will fix it...

**Comment by floryst** [comment]
This is an issue that resulted from #1786, as the default line thickness of 2 makes the line segments extremely large compared to the cone (which has a height of 1 I believe).

With the addition of the line thickness, we need some way of setting the line thickness.
- Add `lineThickness` to the PolyLineWidget's `methodsToLink`
- Have `rep.setLineThickness(...)` set `model.lineThickness`
- Add `rep.getLineThickness()`

This does represent a pretty significant change from what the PolyLineRepresentation was doing before. @finetjul Do you think it's better to separate out the thin polyline rep from the thick polyline rep?

**Comment by finetjul** [comment]
Indeed #1786 breaks the example. Sorry about that.

It would have work if 'scaleInPixels' was true.

I admit it is not clear to me when a property should be set in 'state' (serializable?) and when it should be set in the widget (not serializable?).

While I agree we should add setters and getters on the representation, should the line thickness be set by the example or the widget itself.

> @finetjul Do you think it's better to separate out the thin polyline rep from the thick polyline rep?
I would prefer to keep a unique "polyline representation" class with many options to customize it. The idea behind it is that users (third party applications) should be able to conveniently tweak the rendering to their liking without have to change the class (but maybe I just do not know yet how easy it is to change the representation class)

**Comment by floryst** [comment]
> I admit it is not clear to me when a property should be set in 'state' (serializable?) and when it should be set in the widget (not serializable?).

It's not super clear, but properties that are associated with widget states (e.g. handles) should be bound to widget state, while other properties that can be considered more "global" (e.g. line thickness for a polyline) are bound to the representation. But this distinction definitely needs some discussion.

> I would prefer to keep a unique "polyline representation" class with many options to customize it. The idea behind it is that users (third party applications) should be able to conveniently tweak the rendering to their liking without have to change the class (but maybe I just do not know yet how easy it is to change the representation class)

There is no easy way to change the representation class. Right now you need to extend a widget and modifying the `getRepresentationForViewType` method. We can keep the single polyline representation, but having some means of returning to the original line would be useful (maybe setting lineThickness to 0?)

**Comment by finetjul** [comment]
> It's not super clear, but properties that are associated with widget states (e.g. handles) should be bound to widget state, while other properties that can be considered more "global" (e.g. line thickness for a polyline) are bound to the representation. But this distinction definitely needs some discussion.

I like that distinction. I already moved away "behavior-like" properties from widget state.
Am I right to summarize what you say by stating: "if the "display" property is not meant to change over the lifetime of the widget, then it should not go in the widget state".
In such case, I could see the line thickness being a "dynamic" property. Making the line thicker when a widget is active or something like that (similar to color).

**Comment by floryst** [comment]
After thinking about this, here is another viewpoint when discussing representation properties and widget state. Users can access widget state via `.getWidgetState()` and representation property setters/getters via the widget's `methodsToLink` property. However, I'm pretty sure that the widget's behavior can only access the widget state, so the widget cannot dictate how representations might "look", aka what properties they have.

To sum it up, one approach to deciding between what is a rep property and what is widget state is to see if the widget's behavior needs to modify the property. If there is no need, then maybe that should be left to the user to call `widget.setLineThickness(...)`, assuming that `model.methodsToLink = [..., 'lineThickness', ...]` exists in the widget definition.

**Comment by finetjul** [comment]
I guess that makes sense to me. Thanks for bringing some light on that topic.
